Implement InitFileChunk, ReadFileChunk, and AdvanceFileChunkVersion

Michael-Tu opened the pull request:
Writes will come in the next PR.
Added synchronous unit tests

xicheng87 commented on src/server/chunk_server/chunk_server_file_service_impl.cc:
```diff
@@ -46,8 +137,58 @@ grpc::Status ChunkServerFileServiceImpl::WriteFileChunk(
+        return ConvertProtobufStatusToGrpcStatus(version_or.status());
+      }
+    } else {
+      // NOT FOUND: stale version
```
Ditto. It will be clearer to explicitly state the condition for "stale version", it seems that either `version_or.ValueOrDie() > request->new_chunk_version()` or `version_or.ValueOrDie() < request->new_chunk_version()` can lead to this block, but I think they have different meanings.

xicheng87 commented on src/server/chunk_server/chunk_server_file_service_impl.cc:
```diff
@@ -23,17 +28,103 @@ namespace service {
+      }
+    } else {
+      // NOT FOUND: stale version
+      LOG(ERROR) << "Cannot read file chunk because it is stale: "
```
If we don't distinguish, then it seems FAILED_VERSION_OUT_OF_SYNC is more appropriate to indicate this.

xicheng87 commented on src/server/chunk_server/chunk_server_file_service_impl.cc:
```diff
@@ -23,17 +28,103 @@ namespace service {
     InitFileChunkReply* reply) {
-  // TODO(someone): implement the GFS chunk server logic here
+  LOG(INFO) << "Received InitFileChunkRequest:" << (*request).DebugString();
   *reply->mutable_request() = *request;
```
So far I haven't seen client taking advantage of know the request from the reply message, I presume the original design rationale for this is for async processing.

xicheng87 commented on src/server/chunk_server/file_chunk_manager.cc:
```diff
@@ -102,7 +102,7 @@ google::protobuf::util::StatusOr<std::string> FileChunkManager::ReadFromChunk(
 
   // Check that we aren't trying to read data that isn't there
-  if (start_offset >= file_chunk->data().length()) {
+  if (start_offset > file_chunk->data().length()) {
```
Why eliminating the equal sign here? an offset = data().length() is out of bound right?

xicheng87 commented on tests/server/chunk_server/chunk_server_file_impl_test.cc:
```diff
@@ -0,0 +1,332 @@
+const std::string kTestServerAddress = "0.0.0.0:50053";
+
+// Since we share the same test chunk server in the background, we let differnet
+// unit tests issue requests to *differnet* file handles, so the mutations
```
nit. "different"

xicheng87 commented:
I am approving this change so that we can move faster. Except for the equal sign removal in file_chunk_manager.cc I think other comments are non-essential at this point.

Michael-Tu commented:
Addressed comments by introducing a new enum for out of range, and version out of sync. I don't plan to make it more explicit than is it stale in request or chunk server, to make things easier for now

xicheng87 commented:
LGTM thanks.
